src/main/java/patching/PatchManager.java

```diff
@@ -2,203 +2,433 @@
+        handlingNotificationsLock.readLock().unlock();
 
+        // Immediately send a request if queue non-empty
+        boolean hasNext;
+        synchronized (batchingCtrl.patchBatchingPreQueue) {
+            hasNext = !batchingCtrl.patchBatchingPreQueue.isEmpty();
+        }
         synchronized (batchingCtrl.patchBatchingQueue) {
-            String[] changes = fileChangeNotif.changes;
-
-            for (int i = 0; i < changes.length; i++) {
-                Patch patch = new Patch(changes[i]);
-
-                logger.debug(String.format("PatchManager-Notification: Transforming %s against doneQueue %s", patch, batchingCtrl.patchBatchingQueue).replace("\n", "\\n") + "\n");
-                logger.debug(String.format("PatchManager-Notification: Transforming %s against batchingQueue %s", patch, batchingCtrl.patchDoneQueue).replace("\n", "\\n") + "\n");
-
-                // In event that the
-                Iterator<Patch> itr = batchingCtrl.patchDoneQueue.iterator();
-                while(itr.hasNext()){
-                    Patch donePatch = itr.next();
-                    if(donePatch.getBaseVersion() >= patch.getBaseVersion()){
-                        patch = patch.transform(donePatch);
-                    } else {
-                        // remove ones we don't care about anymore.
-                        itr.remove();
+            hasNext = hasNext || !batchingCtrl.patchBatchingQueue.isEmpty();
+        }
+
+        if (hasNext) {
+            logger.debug("PatchManager-Releaser: Starting next patch-batch");
+            transformAndSendPatch(batchingCtrl, fileID, respHandler, sendErrHandler);
+        } else {
+            logger.debug("PatchManager-Releaser: No next patch-batch; batching and pre-batching queues empty");
+        }
+    }
+
+    // This has to be in a separate thread so that we can have a queue to make sure
+    // notifications are applied in the order they are received. Otherwise the threads waiting for the current
+    // change requests could wake/acquire locks in the wrong order.
+    private void runNotificationHandlerThread() {
+        new Thread(() -> {
+            Thread.currentThread().setName("PatchManagerNotificationHandler");
+            boolean hasWriteLock = false;
+
+            // Loop forever
+            while (true) {
+                Notification notification;
+
+                synchronized (notificationHandlerQueue) {
+                    try {
+                        // Unlock writeLock when notification queue is drained.
+                        if (notificationHandlerQueue.isEmpty()) {
+                            if (hasWriteLock) {
+                                handlingNotificationsLock.writeLock().unlock();
+                                hasWriteLock = false;
+                            }
+                            // Put this thread to sleep while waiting for more notifications.
+                            notificationHandlerQueue.wait();
```

Comment:
It's a wait/notify cycle. It waits so that the synchronization control is unlocked, and is notified by the notificationHandler when items are added into the queue